Expand reference scanning into its own section (#14)
nix-community/nix: doc/manual/src/architecture/store/reference-scanning.md

# Reference scanning

In most cases, Nix can automatically come up with the references that store objects ought to have, saving the user from doing additional bookkeeping.

## Why it works

This might be surprising to users coming from elsewhere!
Traditional build systems do not track dependencies between build artifacts, as the problem appears very challenging since file formats are numerous.
It would seem either there would need to be a way to teach the build system how to parse arbitrary formats, or the build system would need to track references via manual annotations the user must write and, worse, keep up to date.
I
On of the best insights in the creation of Nix was how cut this Gordian knot.

Store objects are, as previously mentioned, mounted at `<store-dir>/<hash>-<name>` when they are exposed to the file system.

> This is the textual form of a [store path](./paths.md).

Reviewer: Isn't the store path the textual form of a store object reference?

Author: In `paths.md` there is a distinction between the conceptual pair of a 20 byte digest and name and the "string representation".

# Reference scanning

As discussed in a previous section, a store object is made up of file system objects and references to other store objects.

Suppose we are trying to create a new store object that *ought* to have references, but we only have plain file system data.
For example, imagine the file system data is a "shared library" (so, dll, dylib) that depends on other shared libraries, and those other shared libraries are already inside other store objects.
How might we do this?

One way would be to have the user manually write down this information.
Another way would be to teach Nix to parse those files' headers and convert the "domain specific" dependency information within them.
Both of these are a huge amount of work, both initially and as ongoing maintenance, and therefore not very attractive options.

Nix instead opts to merely scan the file for store paths without making any attempt to parse it.

> This "simple, stupid" approach actually works quite well most of the time, and imposes very little burden on the end user.

Reviewer: If you say most of the time, you should explain where it does not work. I assume this is [what you intended here](2dd490623fcfc237e63eb99ae70feb6cd1ddd7978df62b53f308bf3cb2e47114R24), but I think it should be there in place. I'm pretty sure Eelco has something very similar very clearly formulated in his thesis.

Author: Examples I have heard of off hand are compressed JAR files and compressed man pages. Are those in the manual?

## Why it works

Store objects are, as previously mentioned, mounted at `<store-dir>/<hash>-<name>` when they are exposed to the file system.
This is the textual form of a [store path](./paths.md).

> When this happens depends on the type of store, but it *must* happen during building, as regular Unix processes are being run.

Reviewer: When what happens? What is "this" here? Reference scanning or mounting?
Also, who runs regular Unix processes? Are there irregular Unix processes?

Author: ```suggestion
When this mounting happens depends on the type of store, but it *must* happen prior to building.
Regular Unix processes that don't necessarily know about Nix are being run, and they need to access the store object inputs of the build.
By mounting the closure of all input store objects, we make them accessible to regular processes as plain directories.
```

Those paths contain hashes, and hashes are not guessable.
As such, build results *must* store those hashes if the programs inside them wish to refer to those store objects without the aid of extra information when they are run.

The hashes could be encrypted or compressed, but regular software doesn't need to do this.
The result is that just scanning for hashes works quite well!

## How it works

> Hashes are scanned for, not entire store paths.

Reviewer: It's not evident to me here why that would be sufficient. Why is the store directory not considered, for example?

Author: I would prefer we scan for the whole thing! I am just documenting the current implementation.

> Thus, nix would look for, e.g.,

Reviewer: ```suggestion
Thus, Nix would look for, e.g.,
```

```
b6gvzjyb2pg0kjfwrjmg1vfhh54ad73z
```
not
```
/nix/store/b6gvzjyb2pg0kjfwrjmg1vfhh54ad73z-firefox-33.1
```

## When it happens

> Scanning happens when store paths are created that could refer to other store paths.

Reviewer: I don't think store paths can refer to anything, as they have no internal structure?
```suggestion
Scanning happens when store objects are created that could refer to other store paths.
```

Reviewer: Do I understand correctly that this sentence is supposed to describe the general case?

Maybe we should somehow first illustrate which ways there are to create store objects in the first place, as we discussed yesterday.

Author: You did fixed it right :)

> When source code is added, references are prohibited by fiat, and thus no scanning is needed.

Reviewer: Interesting. Would be nice to have at least a half-sentence why that is. Is it so we can distinguish sources from everything else, does it have other reasons? Ergonomics?

Author: I don't think there is a good justification for this.

Build results can refer to other objects, so scanning does happen at the end of a build.

## Exceptions

Note that just because Nix *can* scan references doesn't mean that it *must* scan for references.

There is no problem with a store object having an "extra" reference that doesn't correspond to a hash inside the dependency.
The store object will have an extra dependency that it doesn't need, but that is fine.
It is a case adjacent to that of a hash occurring in some obscure location within the store object that is never read.

[Drv files](./drvs/drvs.md), discussed next, might also contain store paths that *aren't* references.
The specific reasons for this will be given then, but we can still ask how how could this possibly be safe?
The fundamental answer is that since the drv file format is known to Nix, it can "do better" than plain scanning.
Nix knows how to parse them, and thus meaningfully differentiate between hashes based on *where* they occur.

> It can decide which contexts correspond to an assumption that the store path ought to exist and be accessible, and which contexts do not.

Reviewer: This seems to be tangential to the mechanism of reference scanning. We haven't introduced derivations as a data structure yet, maybe save that for later.

Author: WIth the "Abstract" vs "Unix" split we discussed, this will be easier :).

# Reference scanning

As discussed in a previous section, a store object is made up of file system objects and references to other store objects.

Suppose we are trying to create a new store object that *ought* to have references, but we only have plain file system data.
For example, imagine the file system data is a "shared library" (so, dll, dylib) that depends on other shared libraries, and those other shared libraries are already inside other store objects.
How might we do this?

> One way would be to have the user manually write down this information.

Reviewer: ```suggestion
One way would be to have the Nix user manually write down this information.
```
